### PR TITLE
Upgrade to tar v3

### PR DESCRIPTION
**Pull request by isaacs**
Tar version 3 performs better and is more well tested than its
predecessor.  npm will be using this in the near future, so there is no
benefit in shipping a node-gyp that uses the slower and less reliable
fstream-based tar.

### CODE REVIEW

File: lib/install.js [hunk 1]
```diff
@@ -218,7 +208,11 @@ function install (gyp, argv, callback) {
         })
 
         // start unzipping and untaring
-        req.pipe(gunzip).pipe(extracter)
+        res.pipe(tar.extract({
+          strip: 1,
+          cwd: devDir,
+          filter: isValid
+        }).on('close', _ => setTimeout(afterTarball, 1000)).on('error', cb))
```
Comment (bnoordhuis) [inline lib/install.js]: Can you use a non-arrow function here?  We still nominally support v0.10 and v0.12.  Does tar@3 support pre-ES6 node?

Side question: why the 1s delay?

Comment (isaacs) [inline lib/install.js]: tar@3 does not support node 0.x.  Neither does npm@5.

I thought they'd fallen off the LTS window.  How long are they going to be nominally supported?

The timeout is a bit of debugging scaffolding that I neglected to remove.  Good catch, thanks.  I'll rebase that out right now.

**Comment by isaacs** [comment]
Note: This drops support for node 0.x, and thus should be considered a breaking semver-major change, and not landed until those node versions are no longer supported.

**Comment by isaacs** [comment]
@bnoordhuis I see at the [LTS schedule](https://github.com/nodejs/LTS) that 0.12 was only in "maintenance" through 2016-12-31.  I think that means it _isn't_ nominally supported?  Still a major version bump, of course.

**Comment by Fishrock123** [comment]
0.12 is no longer supported by Node.js, whether or not 0.12 is still supported by the current version of node-gyp, I am not really sure.

**Comment by addaleax** [comment]
I’d say you can’t just remove 0.x support in a non-major release; too many CI configurations that use `node-gyp` still have those in them.

**Comment by bnoordhuis** [comment]
I suppose that now that node 8 is out, we might as well bump major and drop support for node < 4.  There still seem to be quite a few v0.10.x users but those people probably don't upgrade frequently in general and node-gyp in particular.

**Comment by refack** [comment]
Is there an LST/maintenance plan for `node-gyp`?
FWIW there seem to be [very low code traffic](https://github.com/nodejs/node-gyp/commits/master) and IMHO maintaining a parallel `v3.x` branch should be very easy.

BTW: to the package owners we need a `v3.6.2` for #1206, and #1205

**Comment by refack** [comment]
@isaacs IMHO this PR should come with a `"engines" : { "node" : ">=4" },` clause in `package.json`. Or make this PR dependent on a PR with this change.

**Comment by Fishrock123** [comment]
Ehhhh, if people really want a 3.x maintenance we'll probably only know if they complain about it so I'd say leave it be and consider doing something if people post issues about that specifically.

How do we manage these around here? Merge everything into `master` or another branch or?

**Comment by gibfahn** [comment]
>Ehhhh, if people really want a 3.x maintenance we'll probably only know if they complain about it so I'd say leave it be and consider doing something if people post issues about that specifically.

Agreed

>Merge everything into master or another branch or?

Might as well stay on master, we could create a `3.x` branch if we need to do more 3.x releases going forward.

**Comment by refack** [comment]
> How do we manage these around here? Merge everything into master or another branch or?

I don't know of a president here. I agree with @gibfahn work on `master`, and if needed create a branch from last known good v3, and backport any requested fixes.

**Comment by isaacs** [comment]
You all could probably check the nodejs.org log files to see how many 0.x users are downloading gyp headers, but here at npm, we see that 0.10 and 0.12 together account for a little less than 6% of npm registry traffic, and they're almost all using older versions of npm (and thus, won't get a major version bump of node-gyp).  All other 0.x node versions make up less than 0.1% of npm registry traffic.

In other words, it's almost certainly safe to drop support for 0.x in a major version bump.

**Comment by refack** [comment]
@isaacs if I read this thread correctly there's consensus on dropping < 4 support in a major version bump.

AFAICT We just don't have an established procedure on how to maintain two version lines of a package (only personal experience). IMHO it does not need to be as strict as `node core`'s. Especial since `node-gyp` if very low in code traffic.

I say we just wing it 😉 since collectively we have plenty of prior experience in maintaining packages.

**Comment by isaacs** [comment]
@refack updated package.json engines field to specify >= 4

**Comment by refack** [comment]
> @refack updated package.json engines field to specify >= 4

10x, lets see how much the CI is broken

**Comment by refack** [comment]
https://ci.nodejs.org/job/nodegyp-test-pull-request/21/

**Comment by refack** [comment]
Landed in 5f924ce62c9bca9ab9c2e547bfb87b9a391271ed & 75cfae290fee1791a23fa68820ae5dd841e93e14
